akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/PoolRouterImpl.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.routing
+
+import akka.actor.typed.Behavior.DeferredBehavior
+import akka.actor.typed._
+import akka.actor.typed.scaladsl.AbstractBehavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation.InternalApi
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final case class PoolRouterBuilder[T](
+  poolSize:     Int,
+  behavior:     Behavior[T],
+  logicFactory: () ⇒ RoutingLogic[T] = () ⇒ new RoutingLogics.RoundRobinLogic[T]
+) extends DeferredBehavior[T]
+  with scaladsl.PoolRouter[T]
+  with javadsl.PoolRouter[T] {
+
+  // deferred creation of the actual router
+  def apply(ctx: TypedActorContext[T]): Behavior[T] = PoolRouterImpl[T](poolSize, behavior, logicFactory)
+
+  def withRandomRouting(): PoolRouterBuilder[T] = copy(logicFactory = RoutingLogics.randomLogic[T])
+
+  def withRoundRobinRouting(): PoolRouterBuilder[T] = copy(logicFactory = () ⇒ new RoutingLogics.RoundRobinLogic[T])
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object PoolRouterImpl {
+  def apply[T](poolSize: Int, behavior: Behavior[T], logicFactory: () ⇒ RoutingLogic[T]): Behavior[T] =
+    Behaviors.setup(ctx ⇒ new PoolRouterImpl[T](ctx, poolSize, behavior, logicFactory()))
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private final class PoolRouterImpl[T](
+  ctx:      ActorContext[T],
+  poolSize: Int,
+  behavior: Behavior[T],
+  logic:    RoutingLogic[T]
+) extends AbstractBehavior[T] {
+
+  if (poolSize < 1) throw new IllegalArgumentException(s"pool size must be positive, was $poolSize")
```

Comment:
might be good to add that in `PoolRouterBuilder` too to be able to fail as early as possible